File: services/realtime/internal/auth/ticket.go
// Ticket format mirrors src/shared/lib/realtime/ticket.ts; keep both in sync.
package auth

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

var (
	ErrMalformed = errors.New("malformed ticket")
	ErrSignature = errors.New("invalid ticket signature")
	ErrExpired   = errors.New("expired ticket")
)

type claims struct {
	Sub string `json:"sub"`
	Exp int64  `json:"exp"`
}

func VerifyTicket(ticket string, secret []byte, now time.Time) (string, error) {
	payload, signature, found := strings.Cut(ticket, ".")
	if !found || payload == "" || signature == "" {
		return "", ErrMalformed
	}

	given, err := base64.RawURLEncoding.DecodeString(signature)
	if err != nil {
		return "", ErrMalformed
	}
	if !hmac.Equal(given, sign(payload, secret)) {
		return "", ErrSignature
	}

	raw, err := base64.RawURLEncoding.DecodeString(payload)
	if err != nil {
		return "", ErrMalformed
	}
	var c claims
	if err := json.Unmarshal(raw, &c); err != nil || c.Sub == "" {
		return "", ErrMalformed
	}
	if c.Exp < now.UnixMilli() {
		return "", ErrExpired
	}
	return c.Sub, nil
}

func NewTicket(userID string, secret []byte, expiresAt time.Time) string {
	raw, _ := json.Marshal(claims{Sub: userID, Exp: expiresAt.UnixMilli()})
	payload := base64.RawURLEncoding.EncodeToString(raw)
	return payload + "." + base64.RawURLEncoding.EncodeToString(sign(payload, secret))
}

func sign(payload string, secret []byte) []byte {
	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(payload))
	return mac.Sum(nil)
}

File: src/shared/lib/realtime/ticket.ts
import { createHmac, timingSafeEqual } from 'node:crypto'

export const TICKET_TTL_MS = 60_000

const sign = (payload: string, secret: string) => createHmac('sha256', secret).update(payload).digest('base64url')

/** Format `base64url({ sub, exp }).base64url(hmac-sha256)`, mirrored in services/realtime/internal/auth/ticket.go. */
export function createTicket(userId: string, secret: string, now = Date.now()): string {
  const payload = Buffer.from(JSON.stringify({ sub: userId, exp: now + TICKET_TTL_MS })).toString('base64url')
  return `${payload}.${sign(payload, secret)}`
}

export function verifyTicket(ticket: string | null | undefined, secret: string, now = Date.now()): string | null {
  const [payload, signature] = ticket?.split('.') ?? []
  if (!payload || !signature) return null

  const expected = Buffer.from(sign(payload, secret), 'base64url')
  const given = Buffer.from(signature, 'base64url')
  if (given.length !== expected.length || !timingSafeEqual(given, expected)) return null

  try {
    const { sub, exp } = JSON.parse(Buffer.from(payload, 'base64url').toString('utf8'))
    if (typeof sub !== 'string' || typeof exp !== 'number' || exp < now) return null
    return sub
  } catch {
    return null
  }
}
